### PR TITLE
Adds in-VS migration of xproj projects by shelling out to dotnet migrate.

### PR DESCRIPTION
**Pull request by 333fred**
This addresses #458 and #213. There are still some things to be added to which is tracked by #507, but that depends on the MigrationReport being added to dotnet migrate. These features will be added in follow up PRs at a later date.

Tagging @dotnet/project-system for review.

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.CSharp.VS/ProjectSystem/VS/MigrateXprojProjectFactory.cs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Xproj
+{
+    [Guid(CSharpProjectSystemPackage.XprojTypeGuid)]
+    internal sealed class MigrateXprojProjectFactory : FlavoredProjectFactoryBase, IVsProjectUpgradeViaFactory, IVsProjectUpgradeViaFactory4
+    {
+        public int UpgradeProject(string bstrFileName, uint fUpgradeFlag, string bstrCopyLocation, out string pbstrUpgradedFullyQualifiedFileName, IVsUpgradeLogger pLogger, out int pUpgradeRequired, out Guid pguidNewProjectFactory)
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.CSharp.VS/ProjectSystem/VS/MigrateXprojProjectFactory.cs]: We should probably add a test. It's probably going to need more mocks that it's worth it. Maybe split into smaller method chunks which are unit testable and leave all the VS specific piece out?

Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.CSharp.VS/ProjectSystem/VS/MigrateXprojProjectFactory.cs]: And then add an integration test that migrates some project and verifies the output (once we actually have integration tests running).

File: src/Microsoft.VisualStudio.ProjectSystem.CSharp.VS/ProjectSystem/VS/MigrateXprojProjectFactory.cs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        {
+            var isXproj = bstrFileName.EndsWith(".xproj");
+
+            // If the project is an xproj, then we need to one-way upgrade it. If it isn't, then there's nothing we can do with it.
+            pUpgradeRequired = isXproj ? (int)__VSPPROJECTUPGRADEVIAFACTORYREPAIRFLAGS.VSPUVF_PROJECT_ONEWAYUPGRADE :
+                                         (int)__VSPPROJECTUPGRADEVIAFACTORYREPAIRFLAGS.VSPUVF_PROJECT_NOREPAIR;
+
+            if (isXproj)
+            {
+                pguidNewProjectFactory = new Guid($"{{{CSharpProjectSystemPackage.ProjectTypeGuid}}}");
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.CSharp.VS/ProjectSystem/VS/MigrateXprojProjectFactory.cs]: Double curlies escape the curly in interpolated strings. This is equivalent to `"{" + CSharpProjectSystemPackage.ProjectTypeGuid + "}"`.

**Comment by srivatsn** [comment]
Except for the test that's failing 👍

**Comment by 333fred** [comment]
Yeah, that test is flaky on my machine too. I've converted it to not rely on creating temp files, since it doesn't even need it to begin with.

**Pull request merged**